File: typescript-sdk/src/utils/precision.ts
import { ValidationError } from '../errors.js';

export interface MappedAsset {
    asset_id: string;
    ticker: string;
    precision: number;
    min_order_size: number;
    max_order_size: number;
}

export interface ValidationResult {
    valid: boolean;
    error?: string;
    rawAmount: number;
    minRawAmount: number;
    maxRawAmount: number;
}

export interface OrderSizeLimits {
    minDisplayAmount: number;
    maxDisplayAmount: number;
    minRawAmount: number;
    maxRawAmount: number;
    precision: number;
}

function expandExponentialNumber(value: number): string {
    const str = value.toString();
    if (!/[eE]/.test(str)) {
        return str;
    }

    const match = str.match(/^([+-]?)(\d+)(?:\.(\d+))?[eE]([+-]?\d+)$/);
    if (!match) {
        throw new ValidationError(`Invalid amount: ${str}`);
    }

    const [, sign, integerPart, fractionalPart = '', exponentText] = match;
    const exponent = Number.parseInt(exponentText, 10);
    const digits = integerPart + fractionalPart;
    const decimalIndex = integerPart.length + exponent;

    if (decimalIndex <= 0) {
        return `${sign}0.${'0'.repeat(Math.abs(decimalIndex))}${digits}`;
    }

    if (decimalIndex >= digits.length) {
        return `${sign}${digits}${'0'.repeat(decimalIndex - digits.length)}`;
    }

    return `${sign}${digits.slice(0, decimalIndex)}.${digits.slice(decimalIndex)}`;
}

function expandExponentialString(value: string): string {
    const trimmed = value.trim();
    const match = trimmed.match(/^([+-]?)(?:(\d+)(?:\.(\d*))?|\.(\d+))(?:[eE]([+-]?\d+))?$/);
    if (!match) {
        throw new ValidationError(`Invalid amount: ${value}`);
    }

    const [, sign, integerPartRaw, fractionalPartRaw, leadingFractionRaw, exponentText] = match;
    const integerPart = integerPartRaw ?? '0';
    const fractionalPart = fractionalPartRaw ?? leadingFractionRaw ?? '';
    const digits = `${integerPart}${fractionalPart}`;
    const exponent = exponentText ? Number.parseInt(exponentText, 10) : 0;
    const decimalIndex = integerPart.length + exponent;

    if (/^0+$/.test(digits)) {
        return '0';
    }

    if (decimalIndex <= 0) {
        return `${sign}0.${'0'.repeat(Math.abs(decimalIndex))}${digits}`;
    }

    if (decimalIndex >= digits.length) {
        return `${sign}${digits}${'0'.repeat(decimalIndex - digits.length)}`;
    }

    return `${sign}${digits.slice(0, decimalIndex)}.${digits.slice(decimalIndex)}`;
}

function normalizeDisplayAmount(displayAmount: number, precision: number): string {
    if (!Number.isInteger(precision) || precision < 0) {
        throw new ValidationError(`Precision must be a non-negative integer, got ${precision}`);
    }

    if (!Number.isFinite(displayAmount)) {
        throw new ValidationError(`Amount must be finite, got ${displayAmount}`);
    }

    const normalized = expandExponentialNumber(displayAmount);
    const unsigned =
        normalized.startsWith('-') || normalized.startsWith('+') ? normalized.slice(1) : normalized;
    const [, fractional = ''] = unsigned.split('.');

    if (fractional.length > precision) {
        throw new ValidationError(
            `Amount ${displayAmount} has more than ${precision} decimal places`,
        );
    }

    return normalized;
}

function parseNormalizedDisplayAmount(displayAmount: string | number, precision: number): number {
    const normalized =
        typeof displayAmount === 'number'
            ? normalizeDisplayAmount(displayAmount, precision)
            : (() => {
                  if (!Number.isInteger(precision) || precision < 0) {
                      throw new ValidationError(
                          `Precision must be a non-negative integer, got ${precision}`,
                      );
                  }

                  const exactNormalized = expandExponentialString(displayAmount);
                  const unsigned =
                      exactNormalized.startsWith('-') || exactNormalized.startsWith('+')
                          ? exactNormalized.slice(1)
                          : exactNormalized;
                  const [, fractional = ''] = unsigned.split('.');

                  if (fractional.length > precision) {
                      throw new ValidationError(
                          `Amount ${displayAmount} has more than ${precision} decimal places`,
                      );
                  }

                  return exactNormalized;
              })();
    const sign = normalized.startsWith('-') ? -1 : 1;
    const unsigned =
        normalized.startsWith('-') || normalized.startsWith('+') ? normalized.slice(1) : normalized;
    const [integerPart, fractionalPart = ''] = unsigned.split('.');
    const rawDigits = `${integerPart}${fractionalPart.padEnd(precision, '0')}`;
    const rawAmount = Number(rawDigits || '0') * sign;

    if (!Number.isSafeInteger(rawAmount)) {
        throw new ValidationError(`Raw amount exceeds JavaScript safe integer range`);
    }

    return rawAmount;
}

export class PrecisionHandler {
    private assetPrecisionMap: Map<string, number> = new Map();

    constructor(assets: MappedAsset[]) {
        if (assets.length === 0) {
            throw new Error('Cannot create PrecisionHandler with empty assets array');
        }
        assets.forEach((asset) => {
            this.assetPrecisionMap.set(asset.asset_id, asset.precision);
        });
    }

    toRawAmount(displayAmount: number, assetId: string): number {
        const assetPrecision = this.assetPrecisionMap.get(assetId);
        if (assetPrecision === undefined) {
            throw new Error(`Asset ${assetId} not found in precision handler`);
        }

        return parseRawAmount(displayAmount, assetPrecision);
    }

    toDisplayAmount(rawAmount: number, assetId: string): number {
        const assetPrecision = this.assetPrecisionMap.get(assetId);
        if (assetPrecision === undefined) {
            throw new Error(`Asset ${assetId} not found in precision handler`);
        }

        return rawAmount / Math.pow(10, assetPrecision);
    }

    getAssetPrecision(assetId: string): number {
        const assetPrecision = this.assetPrecisionMap.get(assetId);
        if (assetPrecision === undefined) {
            throw new Error(`Asset ${assetId} not found in precision handler`);
        }
        return assetPrecision;
    }

    formatDisplayAmount(displayAmount: number, assetId: string): string {
        const assetPrecision = this.getAssetPrecision(assetId);
        const normalized = normalizeDisplayAmount(displayAmount, assetPrecision);
        const sign = normalized.startsWith('-') ? '-' : '';
        const unsigned =
            normalized.startsWith('-') || normalized.startsWith('+')
                ? normalized.slice(1)
                : normalized;
        const [integerPart, fractionalPart = ''] = unsigned.split('.');

        return `${sign}${integerPart}.${fractionalPart.padEnd(assetPrecision, '0')}`;
    }

    validateOrderSize(displayAmount: number, asset: MappedAsset): ValidationResult {
        let rawAmount: number;
        try {
            rawAmount = this.toRawAmount(displayAmount, asset.asset_id);
        } catch (error) {
            return {
                valid: false,
                error: error instanceof Error ? error.message : 'Invalid amount',
                rawAmount: 0,
                minRawAmount: asset.min_order_size,
                maxRawAmount: asset.max_order_size,
            };
        }
        const minDisplayAmount = this.toDisplayAmount(asset.min_order_size, asset.asset_id);
        const maxDisplayAmount = this.toDisplayAmount(asset.max_order_size, asset.asset_id);

        if (rawAmount < asset.min_order_size) {
            return {
                valid: false,
                error: `Amount ${displayAmount} ${asset.ticker} is below minimum order size of ${minDisplayAmount} ${asset.ticker}`,
                rawAmount,
                minRawAmount: asset.min_order_size,
                maxRawAmount: asset.max_order_size,
            };
        }

        if (rawAmount > asset.max_order_size) {
            return {
                valid: false,
                error: `Amount ${displayAmount} ${asset.ticker} is above maximum order size of ${maxDisplayAmount} ${asset.ticker}`,
                rawAmount,
                minRawAmount: asset.min_order_size,
                maxRawAmount: asset.max_order_size,
            };
        }

        return {
            valid: true,
            rawAmount,
            minRawAmount: asset.min_order_size,
            maxRawAmount: asset.max_order_size,
        };
    }

    getOrderSizeLimits(asset: MappedAsset): OrderSizeLimits {
        return {
            minDisplayAmount: this.toDisplayAmount(asset.min_order_size, asset.asset_id),
            maxDisplayAmount: this.toDisplayAmount(asset.max_order_size, asset.asset_id),
            minRawAmount: asset.min_order_size,
            maxRawAmount: asset.max_order_size,
            precision: asset.precision,
        };
    }
}

export function createPrecisionHandler(assets: MappedAsset[]): PrecisionHandler {
    return new PrecisionHandler(assets);
}

export function parseRawAmount(displayAmount: string | number, precision: number): number {
    return parseNormalizedDisplayAmount(displayAmount, precision);
}

export function toDisplayAmount(rawAmount: number, precision: number): number {
    if (!Number.isInteger(precision) || precision < 0) {
        throw new ValidationError(`Precision must be a non-negative integer, got ${precision}`);
    }

    return rawAmount / Math.pow(10, precision);
}

File: python-sdk/kaleido_sdk/_utils/precision.py
"""
Precision Handling Utilities

Handles conversion between display amounts (human-readable) and raw amounts (atomic units).
For example: 1.5 BTC (display) <-> 150000000 satoshis (raw, precision=8)
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import ROUND_DOWN, Decimal, InvalidOperation
from typing import TypedDict

from ..errors import ValidationError


class MappedAsset(TypedDict):
    """Asset with precision and order size limits."""

    asset_id: str
    ticker: str
    precision: int
    min_order_size: int
    max_order_size: int


@dataclass
class ValidationResult:
    """Result of order size validation."""

    valid: bool
    raw_amount: int
    min_raw_amount: int
    max_raw_amount: int
    error: str | None = None


@dataclass
class OrderSizeLimits:
    """Order size limits for an asset."""

    min_display_amount: float
    max_display_amount: float
    min_raw_amount: int
    max_raw_amount: int
    precision: int


class PrecisionHandler:
    """
    Precision Handler for converting between display and raw amounts.

    This class manages precision information for multiple assets and provides
    methods to convert between human-readable amounts and atomic units.
    """

    def __init__(self, assets: list[MappedAsset]) -> None:
        """
        Initialize the precision handler with a list of assets.

        Args:
            assets: List of assets with precision information

        Raises:
            ValueError: If assets list is empty
        """
        if not assets:
            raise ValueError("Cannot create PrecisionHandler with empty assets array")

        self._asset_precision_map: dict[str, int] = {}
        for asset in assets:
            self._asset_precision_map[asset["asset_id"]] = asset["precision"]

    def to_raw_amount(self, display_amount: float, asset_id: str) -> int:
        """
        Convert display amount to raw/atomic units.

        Args:
            display_amount: Human-readable amount (e.g., 1.5 BTC)
            asset_id: Asset ID

        Returns:
            Raw amount in atomic units (e.g., 150000000 sats)

        Raises:
            KeyError: If asset not found in precision handler
        """
        precision = self._asset_precision_map.get(asset_id)
        if precision is None:
            raise KeyError(f"Asset {asset_id} not found in precision handler")

        return to_raw_amount(display_amount, precision)

    def to_display_amount(self, raw_amount: int, asset_id: str) -> float:
        """
        Convert raw/atomic units to display amount.

        Args:
            raw_amount: Amount in atomic units (e.g., 150000000 sats)
            asset_id: Asset ID

        Returns:
            Human-readable amount (e.g., 1.5 BTC)

        Raises:
            KeyError: If asset not found in precision handler
        """
        precision = self._asset_precision_map.get(asset_id)
        if precision is None:
            raise KeyError(f"Asset {asset_id} not found in precision handler")

        return to_display_amount(raw_amount, precision)

    def get_asset_precision(self, asset_id: str) -> int:
        """
        Get the precision for an asset.

        Args:
            asset_id: Asset ID

        Returns:
            Number of decimal places

        Raises:
            KeyError: If asset not found in precision handler
        """
        precision = self._asset_precision_map.get(asset_id)
        if precision is None:
            raise KeyError(f"Asset {asset_id} not found in precision handler")
        return precision

    def format_display_amount(self, display_amount: float, asset_id: str) -> str:
        """
        Format a display amount with the correct precision.

        Args:
            display_amount: Amount to format
            asset_id: Asset ID

        Returns:
            Formatted string with correct decimal places
        """
        precision = self.get_asset_precision(asset_id)
        decimal_amount = _normalize_display_amount(display_amount, precision)
        return f"{decimal_amount:.{precision}f}"

    def validate_order_size(self, display_amount: float, asset: MappedAsset) -> ValidationResult:
        """
        Validate that an order size is within the asset's limits.

        Args:
            display_amount: Amount to validate
            asset: Asset with min/max order sizes

        Returns:
            Validation result with error message if invalid
        """
        try:
            raw_amount = self.to_raw_amount(display_amount, asset["asset_id"])
        except ValidationError as exc:
            return ValidationResult(
                valid=False,
                error=str(exc),
                raw_amount=0,
                min_raw_amount=asset["min_order_size"],
                max_raw_amount=asset["max_order_size"],
            )
        min_display = self.to_display_amount(asset["min_order_size"], asset["asset_id"])
        max_display = self.to_display_amount(asset["max_order_size"], asset["asset_id"])

        if raw_amount < asset["min_order_size"]:
            return ValidationResult(
                valid=False,
                error=f"Amount {display_amount} {asset['ticker']} is below minimum order size of {min_display} {asset['ticker']}",
                raw_amount=raw_amount,
                min_raw_amount=asset["min_order_size"],
                max_raw_amount=asset["max_order_size"],
            )

        if raw_amount > asset["max_order_size"]:
            return ValidationResult(
                valid=False,
                error=f"Amount {display_amount} {asset['ticker']} is above maximum order size of {max_display} {asset['ticker']}",
                raw_amount=raw_amount,
                min_raw_amount=asset["min_order_size"],
                max_raw_amount=asset["max_order_size"],
            )

        return ValidationResult(
            valid=True,
            raw_amount=raw_amount,
            min_raw_amount=asset["min_order_size"],
            max_raw_amount=asset["max_order_size"],
        )

    def get_order_size_limits(self, asset: MappedAsset) -> OrderSizeLimits:
        """
        Get the order size limits for an asset.

        Args:
            asset: Asset to get limits for

        Returns:
            Limits in both display and raw formats
        """
        return OrderSizeLimits(
            min_display_amount=self.to_display_amount(asset["min_order_size"], asset["asset_id"]),
            max_display_amount=self.to_display_amount(asset["max_order_size"], asset["asset_id"]),
            min_raw_amount=asset["min_order_size"],
            max_raw_amount=asset["max_order_size"],
            precision=asset["precision"],
        )


def create_precision_handler(assets: list[MappedAsset]) -> PrecisionHandler:
    """
    Factory function to create a PrecisionHandler.

    Args:
        assets: Array of assets with precision information

    Returns:
        New PrecisionHandler instance
    """
    return PrecisionHandler(assets)


def _normalize_display_amount(display_amount: float | Decimal | str, precision: int) -> Decimal:
    if precision < 0:
        raise ValidationError(f"Precision must be non-negative, got {precision}")

    try:
        amount = Decimal(str(display_amount))
    except (InvalidOperation, ValueError) as exc:
        raise ValidationError(f"Invalid amount: {display_amount!r}") from exc

    if not amount.is_finite():
        raise ValidationError(f"Amount must be finite, got {display_amount!r}")

    quantum = Decimal(1).scaleb(-precision)
    normalized = amount.quantize(quantum, rounding=ROUND_DOWN)
    if normalized != amount:
        raise ValidationError(f"Amount {display_amount!r} has more than {precision} decimal places")

    return normalized


def to_raw_amount(display_amount: float, precision: int) -> int:
    """
    Standalone function to convert display amount to raw amount.

    Args:
        display_amount: Human-readable amount
        precision: Number of decimal places

    Returns:
        Raw amount in atomic units
    """
    normalized = _normalize_display_amount(display_amount, precision)
    return int(normalized.scaleb(precision))


def to_display_amount(raw_amount: int, precision: int) -> float:
    """
    Standalone function to convert raw amount to display amount.

    Args:
        raw_amount: Amount in atomic units
        precision: Number of decimal places

    Returns:
        Human-readable amount
    """
    if precision < 0:
        raise ValidationError(f"Precision must be non-negative, got {precision}")

    return float(Decimal(raw_amount).scaleb(-precision))
